### ISSUE TITLE
False Positive: Vec Init Then Push 

### ISSUE THREAD

**Issue by wongjiahau** [rust-lang/rust-clippy#6978]
## Lint name:
[vec_init_then_push](https://rust-lang.github.io/rust-clippy/master/index.html#vec_init_then_push)

## Input
I tried this code:

```rust
let mut type_variables = vec![];
loop {
    type_variables.push(self.eat_token(TokenType::Identifier, context)?);
    if self.try_eat_token(TokenType::MoreThan).is_some() {
        return Ok(type_variables);
    }
}
```

## Expected output
No `vec_init_then_push` lint error. 

## Actual output
```
warning: calls to `push` immediately after creation
   --> src/parse.rs:366:13
    |
366 | /             let mut type_variables = vec![];
367 | |             loop {
368 | |                 type_variables.push(self.eat_token(TokenType::Identifier, context)?);
    | |_____________________________________________________________________________________^ help: consider using the `vec![]` macro: `let mut type_variables = vec![..];`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#vec_init_then_push
```

## Explanation
It should be obvious that the following code snippet (original):
```rust
let mut type_variables = vec![];
loop {
    type_variables.push(self.eat_token(TokenType::Identifier, context)?);
    if self.try_eat_token(TokenType::MoreThan).is_some() {
        return Ok(type_variables);
    }
}
```
behaves differently than(suggested by `clippy`):
```rust
let mut type_variables = vec![
    self.eat_token(TokenType::Identifier, context)?
];
loop {
    if self.try_eat_token(TokenType::MoreThan).is_some() {
        return Ok(type_variables);
    }
}
```
Because the suggested version will only eat one `Identifier`, while the original version will keep on eating `Identifier` as long as the loop is not broken.

### Meta

- `cargo clippy -V`: clippy 0.1.51 (2fd73fab 2021-03-23)
- `rustc -Vv`:
```
rustc 1.51.0 (2fd73fabe 2021-03-23)
binary: rustc
commit-hash: 2fd73fabe469357a12c2c974c140f67e7cdd76d0
commit-date: 2021-03-23
host: x86_64-apple-darwin
release: 1.51.0
LLVM version: 11.0.1
```

**Comment by Y-Nak** [comment]
This seems a duplicate of #6956.

**Comment by wongjiahau** [comment]
@Y-Nak Yes it is, thanks. 

Closing this issue in favour of a more minimal reproducible example.
